spirv_msl.cpp

```diff
@@ -3114,67 +3147,98 @@ void CompilerMSL::add_variable_to_interface_block(StorageClass storage, const st
 		else
 		{
 			bool masked_block = false;
+			uint32_t location = 0;
+			uint32_t var_mbr_idx = 0;
```

Comment:
This is a bug. E.g. this shader emits wrong code:

```
#version 450

struct Foo
{
	float a;
	float b;
};

struct Bar
{
	vec2 a;
	vec2 b;
};

layout(location = 1) in Foo foos[4];
layout(location = 10) in Foo bars[4];
layout(location = 0) out vec4 FragColor;

void main()
{
	FragColor.x = foos[0].a;
	FragColor.y = foos[1].b;
	FragColor.z = foos[2].a;
	FragColor.w = bars[3].b.x;
}
```

Reply:
I assume var_mbr_idx needs to start at size of ib_type.members.

Reply:
The codegen is:
```
    bars[0].a = in.Foo_0_a; <-- alias with foos[] ?
    bars[0].b = in.Foo_0_b;
    bars[1].a = in.Foo_1_a;
    bars[1].b = in.Foo_1_b;
    bars[2].a = in.Foo_2_a;
    bars[2].b = in.Foo_2_b;
    bars[3].a = in.Foo_3_a;
    bars[3].b = in.Foo_3_b;
    out.FragColor.x = foos[0].a;
    out.FragColor.y = foos[1].b;
    out.FragColor.z = foos[2].a;
    out.FragColor.w = bars[3].b;
    return out;
```

Reply:
The problem here is not `var_mbr_idx`, which marks the index into the var struct, not the ib_type struct. 

The core of the problem is that SPIRV-Cross uses **_type_** name to name the ib_type member, so if two vars use the same type, and they both get flattened into the same ib_type, we end up with a name conflict. This is resolved rather arbitrarily in the existing `add_member_name()` to append an arbitrary index when the struct is emitted. Unfortunately, this is too late for the fix-up lambdas that are created to do all the assignment copying above.

I noticed this issue while working on this, and I have a one-line fix in the new code that uses the **_var_** name, instead of the **_var type_** name to derive the ib_type member names. I did not include that change in this PR, because it changes the content of a whole ton of existing reference shaders, which I felt created a lot of review noise for this PR.

With that change in place, the resulting MSL shader becomes that as shown below.

I recommend we pull this PR in, since it's not breaking anything existing, and then I'll do a separate PR with that one-line change, changing all those existing reference shaders, and I'll include your shader above to demonstrate and test for that issue. I recommend doing it that way, because to do it in the existing code would be a larger change (because the existing `to_qualified_member_name()` does not pass in the var).

--------------

```
struct main0_in
{
    float foos_0_a [[user(locn1)]];
    float foos_0_b [[user(locn2)]];
    float foos_1_a [[user(locn3)]];
    float foos_1_b [[user(locn4)]];
    float foos_2_a [[user(locn5)]];
    float foos_2_b [[user(locn6)]];
    float foos_3_a [[user(locn7)]];
    float foos_3_b [[user(locn8)]];
    float bars_0_a [[user(locn10)]];
    float bars_0_b [[user(locn11)]];
    float bars_1_a [[user(locn12)]];
    float bars_1_b [[user(locn13)]];
    float bars_2_a [[user(locn14)]];
    float bars_2_b [[user(locn15)]];
    float bars_3_a [[user(locn16)]];
    float bars_3_b [[user(locn17)]];
};

fragment main0_out main0(main0_in in [[stage_in]])
{
    main0_out out = {};
    spvUnsafeArray<Foo, 4> foos = {};
    spvUnsafeArray<Foo, 4> bars = {};
    foos[0].a = in.foos_0_a;
    foos[0].b = in.foos_0_b;
    foos[1].a = in.foos_1_a;
    foos[1].b = in.foos_1_b;
    foos[2].a = in.foos_2_a;
    foos[2].b = in.foos_2_b;
    foos[3].a = in.foos_3_a;
    foos[3].b = in.foos_3_b;
    bars[0].a = in.bars_0_a;
    bars[0].b = in.bars_0_b;
    bars[1].a = in.bars_1_a;
    bars[1].b = in.bars_1_b;
    bars[2].a = in.bars_2_a;
    bars[2].b = in.bars_2_b;
    bars[3].a = in.bars_3_a;
    bars[3].b = in.bars_3_b;
    out.FragColor.x = foos[0].a;
    out.FragColor.y = foos[1].b;
    out.FragColor.z = foos[2].a;
    out.FragColor.w = bars[3].b;
    return out;
}
```

Reply:
Right. Makes more sense to flatten based on the variable name I guess. Only annoyance is when the SPIRVariable does not have a name, but the type does.